### PR TITLE
RTC driver for PIC32

### PR DESCRIPTION
**Pull request by ghost**
This connects to #58

### CODE REVIEW

File: cpu/pic32/lib/pic32_rtc.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+static uint32_t
+binary_to_bcd(uint32_t a)
+{
+  uint32_t decimal = a / 10;
+  uint32_t digit = a - decimal * 10;
+
+  return (decimal << 4) | digit;
+}
+/*---------------------------------------------------------------------------*/
+int8_t
```
Comment (Avinash95) [inline cpu/pic32/lib/pic32_rtc.c:69]: void ?

Comment (ghost) [reply cpu/pic32/lib/pic32_rtc.c:69]: By default, all public functions return ```int8_t``` to allow returning an error later.

File: cpu/pic32/lib/pic32_rtc.c [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+  return date;
+}
+/*---------------------------------------------------------------------------*/
+int8_t
+pic32_rtc_set_date(struct date date)
+{
+  uint32_t rtccon_state = RTCCON;
+
+  if(date.year > 99
```
Comment (Avinash95) [inline cpu/pic32/lib/pic32_rtc.c]: should we separate conditions of date and time, this if condition is way too big

Comment (ghost) [reply cpu/pic32/lib/pic32_rtc.c]: I think it is more clear if everything is grouped together.

File: cpu/pic32/lib/pic32_rtc.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    uint8_t weekday;  /**< in range 0-6 */
+    uint8_t hour;     /**< in range 0-23 */
+    uint8_t minute;   /**< in range 0-59 */
+    uint8_t second;   /**< in range 0-59 */
+};
+
+int8_t pic32_rtc_init(void);
+struct date pic32_rtc_get_date(void);
+int8_t pic32_rtc_set_date(struct date date);
+
```
Comment (manohar-narkhede-imgtec) [inline cpu/pic32/lib/pic32_rtc.h]: Should we have alarm related functions also?

**Comment by manohar-narkhede-imgtec** [comment]
The test application for RTC testing can be added.

**Comment by ghost** [comment]
I modified hello-world example: 
```
diff --cc examples/hello-world/hello-world.c
index 30724140e1e2,30724140e1e2..a1b6b0f17046
--- a/examples/hello-world/hello-world.c
+++ b/examples/hello-world/hello-world.c
@@@ -40,6 -40,6 +40,32 @@@
  #include "contiki.h"
  
  #include <stdio.h> /* For printf() */
++#include <pic32_rtc.h>
++
++static const char *weekday_str[] = {
++    "Monday",
++    "Tuesday",
++    "Wednesday",
++    "Thursday",
++    "Friday",
++    "Saturday",
++    "Sunday"
++};
++static const char *month_str[] = {
++    "",
++    "January",
++    "Februay",
++    "March",
++    "April",
++    "May",
++    "June",
++    "July",
++    "August",
++    "September",
++    "October",
++    "November",
++    "December"
++};
  /*---------------------------------------------------------------------------*/
  PROCESS(hello_world_process, "Hello world process");
  AUTOSTART_PROCESSES(&hello_world_process);
@@@ -47,9 -47,9 +73,29 @@@
  PROCESS_THREAD(hello_world_process, ev, data)
  {
    PROCESS_BEGIN();
--
--  printf("Hello, world\n");
--  
++  {
++    static struct etimer et;
++    pic32_rtc_init();
++    printf("RTCCON=%08X\n", RTCCON);
++    printf("RTCTIME=%08X\n", RTCTIME);
++    printf("RTCDATE=%08X\n", RTCDATE);
++    struct date initial = {
++      .year = 17,
++      .month = 3,
++      .weekday = 2,
++      .day = 1,
++      .hour = 23,
++      .minute = 12,
++      .second = 54
++    };
++    pic32_rtc_set_date(initial);
++    while (1) {
++      struct date d = pic32_rtc_get_date();
++      printf("%s %u %s %u, %u:%u:%u\n", weekday_str[d.weekday], d.day, month_str[d.month], d.year + 2000, d.hour, d.minute, d.second);
++      etimer_set(&et, CLOCK_SECOND);
++      PROCESS_WAIT_EVENT();
++    }
++  }
    PROCESS_END();
  }
  /*---------------------------------------------------------------------------*/
```

File: cpu/pic32/lib/pic32_rtc.h (hunk 3 above)
Comment (mtusnio) [inline cpu/pic32/lib/pic32_rtc.h]: pic32_rtc_release would be a good addition

File: cpu/pic32/lib/pic32_rtc.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+#ifdef __USE_RTC__
+
+#include <stdint.h>
+
+struct date {
+    uint8_t year;     /**< in range 0..99 */
+    uint8_t month;    /**< in range 1-12 */
+    uint8_t day;      /**< in range 1-31 (depends on month) */
+    uint8_t weekday;  /**< in range 0-6 */
```
Comment (mtusnio) [inline cpu/pic32/lib/pic32_rtc.h]: The weekday range starts from 0, while everything else is numbered from 1 (or 0 as in year's case, but that's for 2000). 1-7 for consistency?

**Review by mtusnio — DISMISSED** [review]

**Comment by manohar-narkhede-imgtec** [comment]
We can not modify the the generic hello-word example. But the above test application can be moved to contiki/platform/mikro-e/apps/test-rtc.

File: cpu/pic32/lib/pic32_rtc.h (hunk 3 above)
Comment (ghost) [reply cpu/pic32/lib/pic32_rtc.h]: Well... it would be better but let's not increase the scope of this pull request. This could be done later.

**Comment by ghost** [comment]
@manohar-narkhede-imgtec I often modify the hello-world example to test features. I posted this patch such that you can easily test this pr.

**Comment by manohar-narkhede-imgtec** [comment]
Then this test app can be added at contiki/platform/mikro-e/apps/test-rtc so that the test application will be there similar to other test applications.

**Review by manohar-narkhede-imgtec — APPROVED** [review]
Verified the test-rtc application.

File: cpu/pic32/lib/pic32_rtc.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+   */
+  RTCCONCLR = _RTCCON_RTCWREN_MASK;
+
+  return 0;
+}
+/*---------------------------------------------------------------------------*/
+int8_t
+pic32_rtc_release(void)
+{
+  SYSKEY = 0;
```
Comment (mtusnio) [inline cpu/pic32/lib/pic32_rtc.c]: Disable interrupts before running the sequence

**Review by mtusnio — CHANGES_REQUESTED** [review]

**Review by mtusnio — APPROVED** [review]

**Pull request merged**